contracts/dtoken/src/rewards.rs

```diff
@@ -0,0 +1,277 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use std::fmt;
+
+#[derive(BorshDeserialize, BorshSerialize, Debug, Serialize, PartialEq, Clone, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+pub enum CampaignType {
+    Supply,
+    Borrow,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Debug, Deserialize, Clone)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Vesting {
+    /// Campaign vesting start time, seconds
+    start_time: u64,
+    /// Campaign vesting end time, seconds
+    end_time: u64,
+    /// Penalty amount which will be arrested in case of early withdraw
+    penalty: Ratio,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Debug, Deserialize, Clone)]
+#[serde(crate = "near_sdk::serde")]
+pub struct RewardCampaign {
+    /// Reward campaign type
+    campaign_type: CampaignType,
+    /// Campaign start time seconds
+    start_time: u64,
+    /// Campaign end time seconds
+    end_time: u64,
+    /// Reward token address
+    token: AccountId,
+    /// Token ticker id
+    ticker_id: String,
+    /// Reward tokens total amount
+    reward_amount: WBalance,
+    /// Last time when rewardPerToken was recomputed/updated
+    last_update_time: u64,
+    /// Represent the token rewards amount which contract should pay for 1 token putted into liquidity
+    rewards_per_token: WBalance,
+    /// Vesting configuration
+    vesting: Vesting,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Debug, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct RewardCampaignExtended {
+    /// Reward campaign id
+    campaign_id: String,
+    /// Reward campaign data
+    campaign: RewardCampaign,
+    /// Market total Supply/Borrow depends on reward campaign type
+    market_total: WBalance,
+    /// Rewards per day token amount
+    rewards_per_day: WBalance,
+}
+
+impl fmt::Display for RewardCampaignExtended {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl Contract {
+    pub fn get_reward_campaign_by_id(&self, campaign_id: String) -> Option<RewardCampaign> {
+        self.reward_campaigns.get(&campaign_id)
+    }
+
+    pub fn get_reward_campaigns_extended(&self) -> Vec<RewardCampaignExtended> {
+        self.reward_campaigns
+            .iter()
+            .filter(|(_, campaign)| campaign.end_time > env::block_height())
+            .map(|(id, campaign)| RewardCampaignExtended {
+                campaign_id: id,
+                campaign: campaign.clone(),
+                market_total: self.get_market_total(campaign.clone()),
+                rewards_per_day: self.get_reward_tokens_per_day(campaign),
+            })
+            .collect::<Vec<RewardCampaignExtended>>()
+    }
+
+    pub fn get_rewards_per_second(&self, campaign: RewardCampaign) -> WBalance {
+        let rewards_per_second: u128 =
+            campaign.reward_amount.0 / u128::from(campaign.end_time - campaign.start_time);
```

Comment:
@pahomas most likely there should be Ratio (or similar type) as we could have something like 0.1 Token per second, which in integer representation will be `0` due to integer division